bin/faucet/src/client.rs

```diff
@@ -0,0 +1,331 @@
+};
+use miden_node_proto::generated::{
+    requests::{GetBlockHeaderByNumberRequest, SubmitProvenTransactionRequest},
+    rpc::api_client::ApiClient,
+};
+use miden_objects::{
+    accounts::{Account, AccountDelta, AccountId, AccountStorageType, AuthSecretKey},
+    assembly::{ModuleAst, ProgramAst},
+    assets::{FungibleAsset, TokenSymbol},
+    crypto::{
+        dsa::rpo_falcon512::SecretKey,
+        merkle::{MmrPeaks, PartialMmr},
+        rand::RpoRandomCoin,
+    },
+    notes::{Note, NoteId, NoteType},
+    transaction::{ChainMmr, ExecutedTransaction, InputNotes, TransactionArgs},
+    vm::AdviceMap,
+    BlockHeader, Felt, Word,
+};
+use miden_tx::{
+    auth::BasicAuthenticator, utils::Serializable, DataStore, DataStoreError, ProvingOptions,
+    TransactionExecutor, TransactionInputs, TransactionProver,
+};
+use rand::{rngs::StdRng, thread_rng, Rng};
+use rand_chacha::{rand_core::SeedableRng, ChaCha20Rng};
+use tonic::transport::Channel;
+
+use crate::{config::FaucetConfig, errors::FaucetError};
+
+pub const DISTRIBUTE_FUNGIBLE_ASSET_SCRIPT: &str =
+    include_str!("transaction_scripts/distribute_fungible_asset.masm");
+
+// FAUCET CLIENT
+// ================================================================================================
+
+/// Basic client that handles execution, proving and submiting of mint transactions
+/// for the faucet.
+pub struct FaucetClient {
+    rpc_api: ApiClient<Channel>,
+    executor: TransactionExecutor<FaucetDataStore, BasicAuthenticator<StdRng>>,
+    data_store: FaucetDataStore,
+    id: AccountId,
+}
+
+unsafe impl Send for FaucetClient {}
+unsafe impl Sync for FaucetClient {}
+
+impl FaucetClient {
+    pub async fn new(config: FaucetConfig) -> Result<Self, FaucetError> {
+        let (rpc_api, root_block_header, root_chain_mmr) =
+            initialize_faucet_client(config.clone()).await?;
+
+        let (faucet_account, account_seed, secret) = build_account(config.clone())?;
+        let faucet_account = Rc::new(RefCell::new(faucet_account));
+        let id = faucet_account.borrow().id();
+
+        let data_store = FaucetDataStore::new(
+            faucet_account.clone(),
+            account_seed,
+            root_block_header,
+            root_chain_mmr,
+        );
+        let authenticator = BasicAuthenticator::<StdRng>::new(&[(
+            secret.public_key().into(),
+            AuthSecretKey::RpoFalcon512(secret),
+        )]);
+        let mut executor =
+            TransactionExecutor::new(data_store.clone(), Some(Rc::new(authenticator)));
+
+        executor
+            .load_account(id)
+            .map_err(|err| FaucetError::InternalServerError(err.to_string()))?;
+
+        Ok(Self { data_store, rpc_api, executor, id })
+    }
+
+    /// Gets [RpoRandomCoin] from the client
+    fn get_random_coin(&self) -> RpoRandomCoin {
+        let mut rng = thread_rng();
+        let coin_seed: [u64; 4] = rng.gen();
+
+        RpoRandomCoin::new(coin_seed.map(Felt::new))
+    }
```

Comment:
Correct me if I'm wrong but I think you added this function because when passing a long-lived RNG object, the object would not be modified and the serial number would be the same for all created notes, right?
@bobbinth is there any reason why [`create_p2id_note()`](https://github.com/0xPolygonMiden/miden-base/blob/727916915ddb9a237ffb975a1f7cc2f9f69139dc/miden-lib/src/notes/mod.rs#L31) does not take `&mut R` instead of owning the object? In that case the faucet could have one `RpoRandomCoin` and pass that every time, no?

Reply:
Yes, I had to change the rng to replicate what the client does to generate random numbers. Befor this fix the serial number was always the same so the note ID didn't change for the same account+asset amout.

Reply:
> @bobbinth is there any reason why [`create_p2id_note()`](https://github.com/0xPolygonMiden/miden-base/blob/727916915ddb9a237ffb975a1f7cc2f9f69139dc/miden-lib/src/notes/mod.rs#L31) does not take `&mut R` instead of owning the object? In that case the faucet could have one `RpoRandomCoin` and pass that every time, no?

hmmm - good catch! There is no good reason why it is `mut R` rather than `&mut R`. @tomyrd - could you make a PR in `miden-base` to fix this and then update this PR accordingly?

Reply:
On it

Reply:
@tomyrd - we should be able to update this now - right?

Reply:
oh - n/m - I see that it is already done. thank you!